packages/experiments/src/index.js

```diff
@@ -0,0 +1,78 @@
+const CORE_MODULES_USING_EXPERIMENTS = [
+	'@wordpress/data',
+	'@wordpress/block-editor',
+	'@wordpress/block-library',
+	'@wordpress/blocks',
+	'@wordpress/core-data',
+	'@wordpress/date',
+	'@wordpress/edit-site',
+	'@wordpress/edit-widgets',
+];
+
+const registeredExperiments = {};
+/*
+ * Warning for theme and plugin developers.
+ *
+ * The use of experimental developer APIs is intended for use by WordPress Core
+ * and the Gutenberg plugin exclusively.
+ *
+ * Dangerously opting in to using these APIs is NOT RECOMMENDED. Furthermore,
+ * the WordPress Core philosophy to strive to maintain backward compatibility
+ * for third-party developers DOES NOT APPLY to experimental APIs.
+ *
+ * THE CONSENT STRING FOR OPTING IN TO THESE APIS MAY CHANGE AT ANY TIME AND
+ * WITHOUT NOTICE. THIS CHANGE WILL BREAK EXISTING THIRD-PARTY CODE. SUCH A
+ * CHANGE MAY OCCUR IN EITHER A MAJOR OR MINOR RELEASE.
+ */
+const requiredConsent =
+	'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.';
+
+export const __dangerousOptInToUnstableAPIsOnlyForCoreModules = (
+	consent,
+	moduleName
+) => {
+	if ( ! CORE_MODULES_USING_EXPERIMENTS.includes( moduleName ) ) {
+		throw new Error(
```

Comment:
Do we want to only throw this message in development build in the future? (Not sure if it's even possible in the current architecture though).

Reply:
No, this adds intentional friction to discourage plugins from even trying.